### PR TITLE
Add regtests for LND, api

### PR DESCRIPTION
**Pull request by Zenith00**
Add method for spawning LND daemon and two btcd backends and a regtest for getexchangerate

### CODE REVIEW

File: routers/apiMethods.js [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+const reqGetAddressDetails = function (args) {
+  return getAddressDetails({
+      address: args.req.params.address,
+      network: args.req.params.network
+    }, returnJson({log: args.log, res: args.res})
+  );
+};
+
+const reqGetExchangeRates = function (args) {
```
Comment (alexbosworth) [inline routers/apiMethods.js]: Style would be to destructure here: `({cache, log, res}) => {}` - args make it harder to see what is being passed. But also functions should have parameters detailed above so you know what you are passing

File: routers/apiMethods.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+module.exports = {
+  reqGetAddressDetails,
+  reqGetExchangeRates,
+  reqGetInvoiceDetails,
+  reqFindSwapOutpoint,
+  reqCreateSwap,
+  reqCheckSwapStatus,
+  reqBroadcastTransaction
+};
```
Comment (alexbosworth) [inline routers/apiMethods.js]: Style would have newline at the end of a file

File: routers/apiMethods.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    }
+  )
+}
+
+module.exports = {
+  reqGetAddressDetails,
+  reqGetExchangeRates,
+  reqGetInvoiceDetails,
+  reqFindSwapOutpoint,
+  reqCreateSwap,
```
Comment (alexbosworth) [inline routers/apiMethods.js]: Style would alphabet these guys

File: test/macros/bitcoincore_type_daemon.js [hunk 4]
```diff
@@ -46,12 +46,15 @@ module.exports = ({dir, network}, cbk) => {
     '-debuglogfile=debug.log',
     '-minrelaytxfee=0',
     '-printtoconsole=1',
-    '-regtest=1',
```
Comment (alexbosworth) [inline test/macros/bitcoincore_type_daemon.js]: Should alphabet these guys

File: test/macros/btcsuite_type_daemon.js [hunk 5]
```diff
@@ -1,43 +1,44 @@
...
+const path = require('path');
 
 const chainServer = require('./../../chain/conf/chain_server_defaults');
 const credentialsForNetwork = require('./../../chain/credentials_for_network');
 const {ECPair} = require('./../../tokenslib');
 const {networks} = require('./../../tokenslib');
 
 const {fromPublicKeyBuffer} = ECPair;
-const knownDaemons = ['btcd', 'ltcd'];
+const knownDaemons = ['btcd', 'ltcd', 'btcdbackend'];
```
Comment (alexbosworth) [inline test/macros/btcsuite_type_daemon.js]: Not sure what this is used for

File: test/macros/btcsuite_type_daemon.js [hunk 6]
```diff
@@ -1,43 +1,44 @@
...
 const notFoundIndex = -1;
 const rpcServerReady = /RPC.server.listening/;
 const unableToStartServer = /Unable.to.start.server/;
 
 /** Start a BTCSuite Type Daemon
 
-  {
```
Comment (alexbosworth) [inline test/macros/btcsuite_type_daemon.js]: No need to change whitespace

File: test/macros/btcsuite_type_daemon.js [hunk 7]
```diff
@@ -1,43 +1,44 @@
...
+   if (knownDaemons.indexOf(args.daemon) === notFoundIndex) {
     return cbk([400, 'ExpectedBtcsuiteDaemonName', args.daemon]);
   }
 
   if (!args.dir) {
     return cbk([400, 'ExpectedDirectoryForDaemon']);
   }
 
-  if (!args.mining_public_key) {
+  if (!args.noMine && !args.mining_public_key && !args.finished_mining_public_key) {
```
Comment (alexbosworth) [inline test/macros/btcsuite_type_daemon.js]: new arguments should be documented above the function and use snake case style

File: test/macros/index.js [hunk 8]
```diff
@@ -8,7 +8,8 @@ const refundSuccess = require('./refund_success');
 const scanForSwap = require('./scan_for_swap');
 const sendChainTokensTransaction = require('./send_chain_tokens_tx');
 const spawnChainDaemon = require('./spawn_chain_daemon');
-
+const spawnLNDDaemon = require('./spawn_lnd_daemon');
```
Comment (alexbosworth) [inline test/macros/index.js]: Style should be aggressive camel case - so spawnLndDaemon

File: test/macros/mock_api.js [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+ @returns
+
+   Response object
+ */
+module.exports = (args, cbk) => {
+  let req = httpMocks.createRequest(args);
+  let res = httpMocks.createResponse();
+  args.method({req, res, log: console, cache: 'memory', cbk});
+
```
Comment (alexbosworth) [inline test/macros/mock_api.js]: Functions should return

File: test/macros/spawn_chain_daemon.js [hunk 10]
```diff
@@ -6,20 +6,20 @@ const bitcoincoreTypeDaemon = require('./bitcoincore_type_daemon')
 const btcsuiteTypeDaemon = require('./btcsuite_type_daemon');
 /** Spawn a chain daemon for testing on regtest
 
-  This method will also listen for uncaught exceptions and stop the daemon
```
Comment (alexbosworth) [inline test/macros/spawn_chain_daemon.js]: No need for white space changes

File: test/macros/spawn_chain_daemon.js [hunk 11]
```diff
@@ -28,9 +28,14 @@ module.exports = (args, cbk) => {
   if (!args.network) {
     return cbk([400, 'ExpectedNetworkTypeForChainDaemon']);
   }
-
+  let dir;
```
Comment (alexbosworth) [inline test/macros/spawn_chain_daemon.js]: Should be able to do a ternary here

File: test/macros/spawn_chain_daemon.js [hunk 12]
```diff
@@ -42,8 +47,13 @@ module.exports = (args, cbk) => {
     daemon = btcsuiteTypeDaemon({
         dir,
         daemon: args.daemon,
+        noMine: args.noMine,
```
Comment (alexbosworth) [inline test/macros/spawn_chain_daemon.js]: Style is to alphabetize the arguments

File: test/macros/spawn_chain_daemon.js [hunk 13]
```diff
@@ -60,23 +70,20 @@ module.exports = (args, cbk) => {
...
-    if (/txn.already.in.mempool/gim.test(data+'')) {
+    if (/txn.already.in.mempool/gim.test(data + '')) {
       return;
     }
-
-    console.log(`${data}`)
   });
 
-  daemon.on('close', code => removeDir(dir, () => {}));
+  // daemon.on('close', code => removeDir(dir, () => {}));
```
Comment (alexbosworth) [inline test/macros/spawn_chain_daemon.js]: Avoid leaving in commented out stuff

File: test/macros/spawn_lnd_daemon.js [hunk 14]
```diff
@@ -0,0 +1,235 @@
+const removeDir = require('rimraf');
+const {spawn} = require('child_process');
+const uuidv4 = require('uuid/v4');
+const asyncAuto = require('async/auto');
+const rpcServerReady = /password RPC server listening on|Finished rescan for 0 addresses/;
```
Comment (alexbosworth) [inline test/macros/spawn_lnd_daemon.js]: Style is alphabetize and put a newline between requires and other var declarations

File: test/macros/spawn_lnd_daemon.js [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+const namespaceFail = /Shutdown complete|Done generating TLS certificates/;
+
+const exec = require('child_process').exec;
+const path = require('path');
+const fs = require('fs');
+const {createAddress} = require('ln-service');
+
+process.env.GRPC_SSL_CIPHER_SUITES = "HIGH+ECDSA";
```
Comment (alexbosworth) [inline test/macros/spawn_lnd_daemon.js]: Better to check on this and make the user set it

File: test/macros/spawn_lnd_daemon.js [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+const {spawn} = require('child_process');
+const uuidv4 = require('uuid/v4');
+const asyncAuto = require('async/auto');
+const rpcServerReady = /password RPC server listening on|Finished rescan for 0 addresses/;
+const namespaceFail = /Shutdown complete|Done generating TLS certificates/;
+
+const exec = require('child_process').exec;
+const path = require('path');
+const fs = require('fs');
+const {createAddress} = require('ln-service');
```
Comment (alexbosworth) [inline test/macros/spawn_lnd_daemon.js]: Style is to separate node module requires from built-in packages - and to use destructuring, and to put built in things and modules up above regular vars

File: test/macros/spawn_lnd_daemon.js [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+
+process.env.GRPC_SSL_CIPHER_SUITES = "HIGH+ECDSA";
+const {lightningDaemon} = require('ln-service');
+const {unlockWallet} = require('ln-service');
+const {getPeers} = require('ln-service');
+const {signMessage} = require('ln-service');
+const returnResult = require('./../../async-util/return_result');
+
+// const {createAddress} = require('ln-service');
```
Comment (alexbosworth) [inline test/macros/spawn_lnd_daemon.js]: Avoid leaving in commented stuff

File: test/macros/spawn_lnd_daemon.js [hunk 18]
```diff
@@ -0,0 +1,235 @@
...
+        chainPass: credentials.pass,
+        chainUser: credentials.user,
+        baseChainPort: credentials.port,
+        lndChainPort: credentials.port - 1,
+      });
+    },
+
+    copyCerts: ['validateCredentials', ({validateCredentials}, cbk) => {
+      fs.mkdirSync(validateCredentials.lndChainDir);
+      fs.copyFile(path.resolve(__dirname, '../swap_regtest', 'dummyrpc.cert'), path.join(validateCredentials.lndChainDir, 'rpc.cert'), copyFail);
```
Comment (alexbosworth) [inline test/macros/spawn_lnd_daemon.js]: Try to keep within 80 col

File: test/macros/spawn_lnd_daemon.js [hunk 19]
```diff
@@ -0,0 +1,235 @@
...
+      const baseChainDir = path.join('/tmp', uuidv4());
+      const lndChainDir = path.join('/tmp', uuidv4());
+
+      return cbk(null, {
+        lndChainDir,
+        baseChainDir,
+        chainPass: credentials.pass,
+        chainUser: credentials.user,
+        baseChainPort: credentials.port,
+        lndChainPort: credentials.port - 1,
```
Comment (alexbosworth) [inline test/macros/spawn_lnd_daemon.js]: Document where this 1 comes from

File: test/macros/spawn_lnd_daemon.js [hunk 20]
```diff
@@ -0,0 +1,235 @@
...
+      }
+      const cert = new Buffer(fs.readFileSync(path.join(spawnLND.lndDir, 'tls.cert'))).toString('base64');
+      const macaroon = new Buffer(fs.readFileSync(path.join(spawnLND.lndDir, 'admin.macaroon'))).toString('base64');
+      const host = '127.0.0.1:10009';
+      const lnd = lightningDaemon({cert, host, macaroon});
+      return cbk(null, {lnd});
+
+    }],
+
+    verifyLNDInterface: ['spawnLNDInterface', ({spawnLNDInterface}, cbk) => {
```
Comment (alexbosworth) [inline test/macros/spawn_lnd_daemon.js]: Try to use camel case in async auto naming

File: test/macros/spawn_lnd_daemon.js [hunk 21]
```diff
@@ -0,0 +1,235 @@
...
+    startNoTLSBTCD: ['genAddress', 'validateCredentials', ({genAddress, validateCredentials}, cbk) => {
+      spawnChainDaemon({
+        network: args.network,
+        daemon: "btcd",
+        dir: validateCredentials.baseChainDir,
+        finished_mining_public_key: genAddress.address,
+        simnet: true,
+        tls: false,
+      }, (err, res) => {
+        if (err) {
```
Comment (alexbosworth) [inline test/macros/spawn_lnd_daemon.js]: Style for single errs is they don't need an else

File: test/swap_regtest/test_get_exchange_rates.js [hunk 22]
```diff
@@ -0,0 +1,19 @@
+const {test} = require('tap');
+const {swapNetworks} = require('./../../chain/conf/api_constants');
+const {mockAPI} = require('../macros');
+const {reqGetExchangeRates} = require('./../../routers/apiMethods');
+
+test(`test exchange rates`, t => {
+  mockAPI({
+    cache: 'memory',
```
Comment (alexbosworth) [inline test/swap_regtest/test_get_exchange_rates.js]: mockApi doesn't seem to take a cache argument

File: chain/conf/api_constants.json [hunk 23]
```diff
@@ -0,0 +1,8 @@
+{
```
[the remark this answers is not in the document]
Comment (Zenith00) [reply chain/conf/api_constants.json]: fixed

File: test/macros/spawn_lnd_daemon.js (hunk 21 above)
Comment (Zenith00) [reply test/macros/spawn_lnd_daemon.js]: fixed

**Pull request closed without merging**